include/deal.II/numerics/vector_tools.templates.h

```diff
@@ -1171,13 +1442,31 @@ namespace VectorTools
                 const Quadrature<dim-1>        &q_boundary,
                 const bool                     project_to_boundary_first)
   {
-    do_project (mapping, dof, constraints, quadrature,
-                function, vec_result,
-                enforce_zero_boundary, q_boundary,
-                project_to_boundary_first);
+    if (dim==spacedim)
+      {
+        const Mapping<dim> *const mapping_ptr = dynamic_cast<const Mapping<dim>*> (&mapping);
```

Comment:
Yes. Constant folding (e.g., the evaluation of `dim==spacedim`) and dead code removal happen at different times. We may get a facility for this in C++17: http://open-std.org/JTC1/SC22/WG21/docs/papers/2016/p0128r1.html

Until then, the idiom used here makes sense to me.